### PR TITLE
Enhance `next dev` performance with placeholder=blur

### PR DESCRIPTION
**Pull request by styfle**
This PR changes the implementation of `placeholder=blur` when using `next dev` so that it lazy loads on-demand.

This will improve the developer experience for web apps with many blurred images.

### CODE REVIEW

**Comment by ijjk** [comment]
# Stats from current PR

<details>
<summary><strong>Default Build</strong> (Decrease detected ✓)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | vercel/next.js enhance-blur-next-dev-lazy | Change |
| - | - | - | - |
| buildDuration | 12.4s | 12.4s | ⚠️ +53ms |
| buildDurationCached | 2.9s | 2.9s | -54ms |
| nodeModulesSize | 49.3 MB | 49.3 MB | ⚠️ +1.72 kB |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall decrease ⚠️</summary>

|  | vercel/next.js canary  | vercel/next.js enhance-blur-next-dev-lazy | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 2.023 | 2.036 | ⚠️ +0.01 |
| / avg req/sec | 1235.65 | 1228.05 | ⚠️ -7.6 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.117 | 1.132 | ⚠️ +0.01 |
| /error-in-render avg req/sec | 2238.31 | 2209.05 | ⚠️ -29.26 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js enhance-blur-next-dev-lazy | Change |
| - | - | - | - |
| 359.HASH.js gzip | 3.09 kB | 3.09 kB | ✓ |
| framework-HASH.js gzip | 42 kB | 42 kB | ✓ |
| main-HASH.js gzip | 20.6 kB | 20.6 kB | ✓ |
| webpack-HASH.js gzip | 1.49 kB | 1.49 kB | ✓ |
| Overall change | 67.2 kB | 67.2 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js enhance-blur-next-dev-lazy | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31.1 kB | 31.1 kB | ✓ |
| Overall change | 31.1 kB | 31.1 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | vercel/next.js enhance-blur-next-dev-lazy | Change |
| - | - | - | - |
| _app-HASH.js gzip | 803 B | 803 B | ✓ |
| _error-HASH.js gzip | 3.18 kB | 3.18 kB | ✓ |
| amp-HASH.js gzip | 526 B | 526 B | ✓ |
| css-HASH.js gzip | 329 B | 329 B | ✓ |
| hooks-HASH.js gzip | 903 B | 903 B | ✓ |
| image-HASH.js gzip | 5.69 kB | 5.69 kB | ✓ |
| index-HASH.js gzip | 261 B | 261 B | ✓ |
| link-HASH.js gzip | 1.66 kB | 1.66 kB | ✓ |
| routerDirect..HASH.js gzip | 319 B | 319 B | ✓ |
| withRouter-HASH.js gzip | 320 B | 320 B | ✓ |
| bb14e60e810b..30f.css gzip | 125 B | 125 B | ✓ |
| Overall change | 14.1 kB | 14.1 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | vercel/next.js enhance-blur-next-dev-lazy | Change |
| - | - | - | - |
| _buildManifest.js gzip | 419 B | 419 B | ✓ |
| Overall change | 419 B | 419 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | vercel/next.js enhance-blur-next-dev-lazy | Change |
| - | - | - | - |
| index.html gzip | 531 B | 531 B | ✓ |
| link.html gzip | 544 B | 544 B | ✓ |
| withRouter.html gzip | 524 B | 524 B | ✓ |
| Overall change | 1.6 kB | 1.6 kB | ✓ |

</details>

</details>
<hr/>
<details>
<summary><strong>Webpack 4 Mode</strong> (Increase detected ⚠️)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | vercel/next.js enhance-blur-next-dev-lazy | Change |
| - | - | - | - |
| buildDuration | 10s | 10s | ⚠️ +17ms |
| buildDurationCached | 4s | 3.9s | -76ms |
| nodeModulesSize | 49.3 MB | 49.3 MB | ⚠️ +1.72 kB |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall increase ✓</summary>

|  | vercel/next.js canary  | vercel/next.js enhance-blur-next-dev-lazy | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 2.062 | 2.061 | 0 |
| / avg req/sec | 1212.39 | 1212.72 | +0.33 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.138 | 1.093 | -0.04 |
| /error-in-render avg req/sec | 2197.55 | 2288.16 | +90.61 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js enhance-blur-next-dev-lazy | Change |
| - | - | - | - |
| 14.HASH.js gzip | 3.11 kB | 3.11 kB | ✓ |
| 677f882d2ed8..HASH.js gzip | 13.9 kB | 13.9 kB | ✓ |
| framework.HASH.js gzip | 41.8 kB | 41.8 kB | ✓ |
| main-HASH.js gzip | 7.81 kB | 7.81 kB | ✓ |
| webpack-HASH.js gzip | 1.19 kB | 1.19 kB | ✓ |
| Overall change | 67.8 kB | 67.8 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js enhance-blur-next-dev-lazy | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31.3 kB | 31.3 kB | ✓ |
| Overall change | 31.3 kB | 31.3 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | vercel/next.js enhance-blur-next-dev-lazy | Change |
| - | - | - | - |
| _app-HASH.js gzip | 791 B | 791 B | ✓ |
| _error-HASH.js gzip | 3.83 kB | 3.83 kB | ✓ |
| amp-HASH.js gzip | 531 B | 531 B | ✓ |
| css-HASH.js gzip | 333 B | 333 B | ✓ |
| hooks-HASH.js gzip | 910 B | 910 B | ✓ |
| index-HASH.js gzip | 230 B | 230 B | ✓ |
| link-HASH.js gzip | 1.64 kB | 1.64 kB | ✓ |
| routerDirect..HASH.js gzip | 297 B | 297 B | ✓ |
| withRouter-HASH.js gzip | 293 B | 293 B | ✓ |
| e025d2764813..52f.css gzip | 125 B | 125 B | ✓ |
| Overall change | 8.98 kB | 8.98 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | vercel/next.js enhance-blur-next-dev-lazy | Change |
| - | - | - | - |
| _buildManifest.js gzip | 418 B | 418 B | ✓ |
| Overall change | 418 B | 418 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | vercel/next.js enhance-blur-next-dev-lazy | Change |
| - | - | - | - |
| index.html gzip | 574 B | 574 B | ✓ |
| link.html gzip | 586 B | 586 B | ✓ |
| withRouter.html gzip | 569 B | 569 B | ✓ |
| Overall change | 1.73 kB | 1.73 kB | ✓ |

</details>

</details>
Commit: 0059e0691e37c7c1a38dc3b74b9d99e87c77af56

**Comment by ijjk** [comment]
## Failing test suites 
Commit: 0059e0691e37c7c1a38dc3b74b9d99e87c77af56

**test/integration/config-devtool-dev/test/index.test.js** 
- devtool set in development mode in next config > should warn and revert when a devtool is set in development mode

<details>
<summary>Expand output</summary>

  ● devtool set in development mode in next config › should warn and revert when a devtool is set in development mode

    thrown: "Exceeded timeout of 30000 ms for a test.
    Use jest.setTimeout(newTimeout) to increase the timeout value, if this is a long-running test."

      17 |
      18 | describe('devtool set in development mode in next config', () => {
    > 19 |   it('should warn and revert when a devtool is set in development mode', async () => {
         |   ^
      20 |     let stderr = ''
      21 |
      22 |     const appPort = await findPort()

      at integration/config-devtool-dev/test/index.test.js:19:3
      at Object.<anonymous> (integration/config-devtool-dev/test/index.test.js:18:1)

</details>

File: packages/next/build/webpack-config.ts [hunk 1]
```diff
@@ -1034,6 +1034,8 @@ export default async function getBaseWebpackConfig(
                 dependency: { not: ['url'] },
                 options: {
                   isServer,
+                  isDev: dev,
+                  configImageSizes: config.images.imageSizes,
```
Comment (timneutkens) [inline packages/next/build/webpack-config.ts:1038]: Doesn't seem like this line is used in the loader

File: packages/next/build/webpack/loaders/next-image-loader.js [hunk 2]
```diff
@@ -9,14 +9,15 @@ const VALID_BLUR_EXT = ['jpeg', 'png', 'webp']
...
-    const isServer = loaderUtils.getOptions(this).isServer
+    const { isServer, isDev } = loaderUtils.getOptions(this)
     const context = this.rootContext
     const opts = { context, content }
     const interpolatedName = loaderUtils.interpolateName(
       this,
       '/static/image/[path][name].[hash].[ext]',
       opts
     )
+    const outputPath = '/_next' + interpolatedName
```
Comment (timneutkens) [inline packages/next/build/webpack/loaders/next-image-loader.js:20]: This needs to consider `assetPrefix`

File: packages/next/build/webpack-config.ts (hunk 1 above)
Comment (styfle) [reply packages/next/build/webpack-config.ts:1038]: Nice catch, I’ll remove it. 

Originally I was thinking of using the smallest size defined in the config but then realized that will yield very different results depending on the config so I just special cases the API instead.

File: packages/next/build/webpack/loaders/next-image-loader.js (hunk 2 above)
Comment (styfle) [reply packages/next/build/webpack/loaders/next-image-loader.js:20]: I see, the `assetPrefix` is only needed for the URL, not the file system path, right?

**Comment by ijjk** [comment]
# Stats from current PR

<details>
<summary><strong>Default Build</strong> (Decrease detected ✓)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | vercel/next.js enhance-blur-next-dev-lazy | Change |
| - | - | - | - |
| buildDuration | 16s | 15.8s | -207ms |
| buildDurationCached | 3.7s | 3.6s | -115ms |
| nodeModulesSize | 49.3 MB | 49.3 MB | ⚠️ +1.83 kB |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall decrease ⚠️</summary>

|  | vercel/next.js canary  | vercel/next.js enhance-blur-next-dev-lazy | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 2.573 | 2.58 | ⚠️ +0.01 |
| / avg req/sec | 971.65 | 968.89 | ⚠️ -2.76 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.451 | 1.505 | ⚠️ +0.05 |
| /error-in-render avg req/sec | 1723.21 | 1661.66 | ⚠️ -61.55 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js enhance-blur-next-dev-lazy | Change |
| - | - | - | - |
| 359.HASH.js gzip | 3.09 kB | 3.09 kB | ✓ |
| framework-HASH.js gzip | 42 kB | 42 kB | ✓ |
| main-HASH.js gzip | 20.6 kB | 20.6 kB | ✓ |
| webpack-HASH.js gzip | 1.49 kB | 1.49 kB | ✓ |
| Overall change | 67.2 kB | 67.2 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js enhance-blur-next-dev-lazy | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31.1 kB | 31.1 kB | ✓ |
| Overall change | 31.1 kB | 31.1 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | vercel/next.js enhance-blur-next-dev-lazy | Change |
| - | - | - | - |
| _app-HASH.js gzip | 803 B | 803 B | ✓ |
| _error-HASH.js gzip | 3.18 kB | 3.18 kB | ✓ |
| amp-HASH.js gzip | 526 B | 526 B | ✓ |
| css-HASH.js gzip | 329 B | 329 B | ✓ |
| hooks-HASH.js gzip | 903 B | 903 B | ✓ |
| image-HASH.js gzip | 5.69 kB | 5.69 kB | ✓ |
| index-HASH.js gzip | 261 B | 261 B | ✓ |
| link-HASH.js gzip | 1.66 kB | 1.66 kB | ✓ |
| routerDirect..HASH.js gzip | 319 B | 319 B | ✓ |
| withRouter-HASH.js gzip | 320 B | 320 B | ✓ |
| bb14e60e810b..30f.css gzip | 125 B | 125 B | ✓ |
| Overall change | 14.1 kB | 14.1 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | vercel/next.js enhance-blur-next-dev-lazy | Change |
| - | - | - | - |
| _buildManifest.js gzip | 419 B | 419 B | ✓ |
| Overall change | 419 B | 419 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | vercel/next.js enhance-blur-next-dev-lazy | Change |
| - | - | - | - |
| index.html gzip | 531 B | 531 B | ✓ |
| link.html gzip | 544 B | 544 B | ✓ |
| withRouter.html gzip | 524 B | 524 B | ✓ |
| Overall change | 1.6 kB | 1.6 kB | ✓ |

</details>

</details>
<hr/>
<details>
<summary><strong>Webpack 4 Mode</strong> (Decrease detected ✓)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | vercel/next.js enhance-blur-next-dev-lazy | Change |
| - | - | - | - |
| buildDuration | 12.6s | 12.7s | ⚠️ +107ms |
| buildDurationCached | 5s | 4.9s | -46ms |
| nodeModulesSize | 49.3 MB | 49.3 MB | ⚠️ +1.83 kB |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall decrease ⚠️</summary>

|  | vercel/next.js canary  | vercel/next.js enhance-blur-next-dev-lazy | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 2.579 | 2.567 | -0.01 |
| / avg req/sec | 969.22 | 973.77 | +4.55 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.397 | 1.48 | ⚠️ +0.08 |
| /error-in-render avg req/sec | 1789.01 | 1689.11 | ⚠️ -99.9 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js enhance-blur-next-dev-lazy | Change |
| - | - | - | - |
| 14.HASH.js gzip | 3.11 kB | 3.11 kB | ✓ |
| 677f882d2ed8..HASH.js gzip | 13.9 kB | 13.9 kB | ✓ |
| framework.HASH.js gzip | 41.8 kB | 41.8 kB | ✓ |
| main-HASH.js gzip | 7.81 kB | 7.81 kB | ✓ |
| webpack-HASH.js gzip | 1.19 kB | 1.19 kB | ✓ |
| Overall change | 67.8 kB | 67.8 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js enhance-blur-next-dev-lazy | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31.3 kB | 31.3 kB | ✓ |
| Overall change | 31.3 kB | 31.3 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | vercel/next.js enhance-blur-next-dev-lazy | Change |
| - | - | - | - |
| _app-HASH.js gzip | 791 B | 791 B | ✓ |
| _error-HASH.js gzip | 3.83 kB | 3.83 kB | ✓ |
| amp-HASH.js gzip | 531 B | 531 B | ✓ |
| css-HASH.js gzip | 333 B | 333 B | ✓ |
| hooks-HASH.js gzip | 910 B | 910 B | ✓ |
| index-HASH.js gzip | 230 B | 230 B | ✓ |
| link-HASH.js gzip | 1.64 kB | 1.64 kB | ✓ |
| routerDirect..HASH.js gzip | 297 B | 297 B | ✓ |
| withRouter-HASH.js gzip | 293 B | 293 B | ✓ |
| e025d2764813..52f.css gzip | 125 B | 125 B | ✓ |
| Overall change | 8.98 kB | 8.98 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | vercel/next.js enhance-blur-next-dev-lazy | Change |
| - | - | - | - |
| _buildManifest.js gzip | 418 B | 418 B | ✓ |
| Overall change | 418 B | 418 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | vercel/next.js enhance-blur-next-dev-lazy | Change |
| - | - | - | - |
| index.html gzip | 574 B | 574 B | ✓ |
| link.html gzip | 586 B | 586 B | ✓ |
| withRouter.html gzip | 569 B | 569 B | ✓ |
| Overall change | 1.73 kB | 1.73 kB | ✓ |

</details>

</details>
Commit: f79cf91da225278b7e2fd7226a8bb93821f0366c

File: packages/next/build/webpack/loaders/next-image-loader.js (hunk 2 above)
Comment (timneutkens) [reply packages/next/build/webpack/loaders/next-image-loader.js:20]: Yeah indeed!

**Review by timneutkens — APPROVED** [review]
Looks good, can you follow up with a PR to add a test for assetPrefix when using static image

**Pull request merged**